### PR TITLE
[FLINK-23728][state-processor-api] State bootstrapping fails on new state backend factory stack

### PR DESCRIPTION
**Pull request by sjwiesman**
## What is the purpose of the change

Support the new state backend interfaces in the state processor api by forcing checkpoint storage to always use FileSystemCheckpointStorage.

## Brief change log

See above

## Verifying this change

Added new state backends to the test suite. 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / **docs** / JavaDocs / not documented)

### CODE REVIEW

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java [hunk 1]
```diff
@@ -18,15 +18,14 @@
...
+import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend;
 
 /** IT Case for reading state from a RocksDB keyed state backend. */
 public class RocksDBStateBackendReaderKeyedStateITCase
-        extends SavepointReaderKeyedStateITCase<RocksDBStateBackend> {
+        extends SavepointReaderKeyedStateITCase<EmbeddedRocksDBStateBackend> {
+
     @Override
-    protected RocksDBStateBackend getStateBackend() {
-        return new RocksDBStateBackend((StateBackend) new MemoryStateBackend());
```
Comment (NicoK) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendReaderKeyedStateITCase.java:30]: same here...
(Is the old way of configuring these still working?)

File: flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendWindowITCase.java [hunk 2]
```diff
@@ -18,16 +18,14 @@
...
+import org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend;
 
 /** IT Case for reading window state with the memory state backend. */
 public class RocksDBStateBackendWindowITCase
-        extends SavepointWindowReaderITCase<RocksDBStateBackend> {
+        extends SavepointWindowReaderITCase<EmbeddedRocksDBStateBackend> {
 
     @Override
-    protected RocksDBStateBackend getStateBackend() {
-        return new RocksDBStateBackend((StateBackend) new MemoryStateBackend());
```
Comment (NicoK) [inline flink-libraries/flink-state-processing-api/src/test/java/org/apache/flink/state/api/RocksDBStateBackendWindowITCase.java:31]: same here...

**Review by NicoK — COMMENTED** [review]
A few details are a bit unclear to me (I left comments) and I'm a bit worried about the test coverage for the old state backends for the savepoint reader ITCases

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java [hunk 3]
```diff
@@ -215,12 +217,23 @@ StreamConfig getConfig(
...
     }
 
+    /**
+     * @return A checkpoint storage that will write the resulting savepoint to a DFS. It is
+     *     configured with a dummy checkpoint directory to satisfy the API but should only ever be
+     *     used for savepoints. This is required to override the default checkpoint directory which
+     *     is always in memory.
+     */
+    private static CheckpointStorage getCheckpointStorage() {
+        return new FileSystemCheckpointStorage("file:///tmp/unused");
```
[the remark this answers is not in the document]
Comment (sjwiesman) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java:234]: yeah, that's kind of where I'm at. There aren't any particularly great solutions available here. I will explicitly disable checkpointing in the stream config for added safety.

**Comment by sjwiesman** [comment]
@NicoK I've re-added the tests you mentioned. 

Regarding the main commit: The test is not great because it only arises when the snapshot is > 5mb. I can artificially increase the amount of data stored in the snapshot, but I'm concerned about that creating test instabilities. If anyone has any ideas please lmk.

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (NicoK) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java:235]: Can you actually do that?
A couple of lines above the new code is this:
```
        config.setCheckpointingEnabled(true);
        config.setCheckpointMode(CheckpointingMode.EXACTLY_ONCE);
```

Another thought: the only place where this method is called is inside `writeOperatorSubtaskStates()`. What do you think about using some "dummy" path beneath the `savepointPath` that is known there?
I mean, with the additional safeguard, it shouldn't be used anyway, but that way, it's at least connected to the input parameters and would be seen by the user if it breaks...

[the remark this answers is not in the document]
Comment (rkhachatryan) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java]: What do you think about of 
```
    private static CheckpointStorage getCheckpointStorage() {
        return new CheckpointStorage() {
            @Override
            public CompletedCheckpointStorageLocation resolveCheckpoint(String externalPointer)
                    throws IOException {
                throw new UnsupportedOperationException();
            }

            @Override
            public CheckpointStorageAccess createCheckpointStorage(JobID jobId) throws IOException {
                return new org.apache.flink.streaming.api.operators.sorted.state.NonCheckpointingStorageAccess();
            }
        };
    }
```
?

(`NonCheckpointingStorageAccess` needs to be made `public` and maybe moved/renamed).

**Review by NicoK — APPROVED** [review]
lgtm

File: flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sjwiesman) [reply flink-libraries/flink-state-processing-api/src/main/java/org/apache/flink/state/api/BootstrapTransformation.java]: I will avoid that for now. That's a bigger refactor then I want to make at this time.

**Pull request closed without merging**